### PR TITLE
Fixes for partitioner with symbolic shapes

### PR DESCRIPTION
**Pull request by wconstab**
- supports saving symint (and symfloat..) values between fw/bwd, using sketchy logic that probably needs to be improved but seems to work so far
- sets a correct weight=1 for sym nodes for cost purposes
- lets user functions return symints/floats (but if the same symfloat is saved for backward, that gets duplicated annoyingly)
- makes partitioning decisions based on observed trace-time sizes without guarding! (this is sketchy, but it isn't clear that it will lead to bad partitioning choices either)
- improves infra for tracking symint-family of types: is_sym_node() and _py_sym_types

### CODE REVIEW

File: functorch/_src/aot_autograd.py
Comment (ezyang) [inline functorch/_src/aot_autograd.py:391]: @Chillee I hate that there is no built in API for getting the unique(!) output node lol

Comment (wconstab) [reply functorch/_src/aot_autograd.py:391]: yea we should add that

File: functorch/test/test_aotdispatch.py
Comment (wconstab) [inline functorch/test/test_aotdispatch.py:583]: clean this up.

File: functorch/test/test_aotdispatch.py
Comment (wconstab) [inline functorch/test/test_aotdispatch.py:628]: thoughts here?

File: torch/fx/experimental/proxy_tensor.py
Comment (ezyang) [inline torch/fx/experimental/proxy_tensor.py:60]: Our tracing invariant is that ALL nodes produced by post-functionalization tracing have a val attached to them. Pre-functionalization tracing we cannot guarantee this for tensors as there is no way to preserve both object identity and metadata mutation in the case you have something like `x = y.resize_(blah)` in the graph (x and y are the same object, but their metadata changes depending on if you're before or after the line. Because we only run this in the partitioner post functionalization there is no problem. However, if you are afraid of this case, we can also introduce an alternate field for meta which just says what the kind of node it is (Tensor or SymInt/SymFloat)

File: torch/fx/experimental/proxy_tensor.py
Comment (ezyang) [inline torch/fx/experimental/proxy_tensor.py:63]: I think you should assert that these fields exist and error if they do not.

File: functorch/_src/aot_autograd.py
Comment (ezyang) [inline functorch/_src/aot_autograd.py:422]: What guarantees that the symints are the suffix of the graph in question?

File: functorch/_src/aot_autograd.py
Comment (ezyang) [inline functorch/_src/aot_autograd.py:395]: If the symints are guaranteed to be suffix, we can assert that this is the case here.

File: functorch/_src/aot_autograd.py
Comment (ezyang) [inline functorch/_src/aot_autograd.py:421]: You don't actually need this condition, I think

File: functorch/_src/aot_autograd.py
Comment (ezyang) [inline functorch/_src/aot_autograd.py:434]: @bdhirsh this looks like this would fix the None problem

**Review by ezyang — APPROVED** [review]
Assuming the invariants are right, this looks correct to me. Good job. Probably should wait for a Horace review too

File: functorch/_src/aot_autograd.py
Comment (wconstab) [reply functorch/_src/aot_autograd.py:422]: this is one of the gaps i need to figure this out better before i land it. i couldn't figure out a test case that violated this, but i also didn't prove it's true.

Comment (ezyang) [reply functorch/_src/aot_autograd.py:422]: The graph you are partitioning here is generated by joint graph in FX. So that's the code to audit. But I don't see why it would organize the outputs this way. Maybe the sizes are getting shunted to the end because of lazy tracing, but if so, that's a very delicate way the invariant is coming up and we will need to handle it explicitly. Maybe @Chillee knows.

File: functorch/test/test_aotdispatch.py
Comment (ezyang) [reply functorch/test/test_aotdispatch.py:628]: It'd be easier to say if you pasted the graph here.

File: functorch/_src/aot_autograd.py
Comment (wconstab) [reply functorch/_src/aot_autograd.py:421]: haha i do. but i want to rewrite this logic anyway.  i did not know this, but it turns out that `something[start:-0]` is not the same as `something[start:]`

Comment (ezyang) [reply functorch/_src/aot_autograd.py:421]: lol

File: torch/fx/experimental/proxy_tensor.py
Comment (wconstab) [reply torch/fx/experimental/proxy_tensor.py]: the case where this breaks down is tuples.  example:
```
node.op ->"call_function"
node.target -> <OpOverload(op='aten.native_batch_norm', overload='default')>
"val" in node.meta: False
node.users -> {getitem: None, getitem_1: None, getitem_2: None}
```

i wonder if we can tag this case during set_meta, using anything other than "all users are getitem' heuristic.  Or maybe that heuristic is good enough?

File: functorch/_src/aot_autograd.py
Comment (Chillee) [reply functorch/_src/aot_autograd.py:422]: @wconstab No I don't think it's true that they're always at the end. They should be ordered by their relative position in the original graph: https://github.com/pytorch/pytorch/blob/master/functorch/_src/partitioners.py#L392-L393

Although activations are always after the original outputs.

File: torch/fx/experimental/proxy_tensor.py
Comment (Chillee) [reply torch/fx/experimental/proxy_tensor.py]: We would just do it before we trace the `getitem` unpacking: https://github.com/pytorch/pytorch/blob/master/torch/fx/experimental/proxy_tensor.py#L161

so
```
    if isinstance(inner_res, tuple) or isinstance(inner_res, list):
        (+) proxy_res.node.meta['val'] = inner_res (or maybe something else, dunno)
```

File: functorch/test/test_aotdispatch.py
Comment (wconstab) [reply functorch/test/test_aotdispatch.py]: fixed this i think

File: functorch/_src/aot_autograd.py
Comment (wconstab) [reply functorch/_src/aot_autograd.py:422]: i think i fixed this now. pls take a look

Comment (ezyang) [reply functorch/_src/aot_autograd.py:422]: The logic... looks the same? What am I looking at

File: functorch/_src/aot_autograd.py
Comment (wconstab) [reply functorch/_src/aot_autograd.py:395]: TODO assert this.

File: functorch/_src/partitioners.py
Comment (wconstab) [inline functorch/_src/partitioners.py:174]: @ezyang @Chillee this is 1/2 important change: separate lists of saved tensors vs saved sym nodes

File: functorch/_src/partitioners.py
Comment (wconstab) [inline functorch/_src/partitioners.py:95]: @ezyang @Chillee this is 2/2 where the saved sym nodes are separately handled, pruned out for unused ones, and then applied to the fwd/bwd args

as i write this, i wonder why i don't need to reverse the order of the contents of saved_sym_nodes as i place it at the beginning of bwd args.

File: functorch/_src/aot_autograd.py
Comment (ezyang) [reply functorch/_src/aot_autograd.py:421]: you could fix it with `something[start:len(something)-num_symints]`

File: functorch/_src/partitioners.py
Comment (ezyang) [inline functorch/_src/partitioners.py:172]: why not just directly put things into `saved_value` or `saved_sym_nodes` in the initial loop

**Comment by ezyang** [comment]
new changes look correct, but it's very delicate so maybe more comments would be good

File: functorch/_src/aot_autograd.py
Comment (bdhirsh) [reply functorch/_src/aot_autograd.py:433]: it did :)

File: torch/fx/experimental/proxy_tensor.py
Comment (wconstab) [inline torch/fx/experimental/proxy_tensor.py:58]: @SherlockNoMad i'll land this soon and maybe you want to use this in your PR

**Pull request closed without merging**
